extension/source/background/KeyringController.ts

```diff
@@ -154,6 +158,76 @@ export default class KeyringController {
 
       await this.keyring.update({ wallets: newWallets });
     },
+
+    /**
+     * Recovers an existing BLS wallet and adds the new
+     * recovered wallet to the keyring
+     * @param recoveryWalletAddress Smart contract address
+     * of wallet being recovered
+     * @param recoverySaltHash Salt used to set the recovery
+     * hash on the wallet that is being recovered
+     */
+    addRecoveryWallet: async ({
+      params: [recoveryWalletAddress, recoverySaltHash],
+    }) => {
+      const netCfg = getNetworkConfig(
+        await this.network.read(),
+        this.multiNetworkConfig,
+      );
+
+      // eslint-disable-next-line camelcase
+      const verificationGatewayContract = VerificationGateway__factory.connect(
+        netCfg.addresses.verificationGateway,
+        await this.ethersProvider.read(),
+      );
+
+      const recoveryWalletHash =
+        await verificationGatewayContract.hashFromWallet(recoveryWalletAddress);
+
+      // Create new private key for the wallet we are recovering to.
+      const newPrivateKey = generateRandomHex(256);
+
+      const addressSignature = await this.signWalletAddress(
+        recoveryWalletAddress,
+        newPrivateKey,
+      );
+
+      // Get instance of the new wallet, so we can get the public key
+      // to pass to the recoverWallet method.
+      const newWalletWrapper = await this.BlsWalletWrapper(newPrivateKey);
+
+      const signerPublicKeyHash = await this.selectedPublicKeyHash.read();
+
+      assert(
+        signerPublicKeyHash !== undefined,
+        () => new Error('Selected public key hash not found'),
+      );
+
+      const tx = {
+        from: signerPublicKeyHash,
+        to: verificationGatewayContract.address,
+        value: '0',
+        data: verificationGatewayContract.interface.encodeFunctionData(
+          'recoverWallet',
+          [
+            addressSignature,
+            recoveryWalletHash,
+            recoverySaltHash,
+            newWalletWrapper.PublicKey(),
+          ],
+        ),
+      };
+
+      this.InternalRpc().eth_sendTransaction({
+        ...tx,
+        gas: undefined,
+        gasPrice: undefined,
+        data: undefined,
+      });
+
+      // Add new private key
+      await this.InternalRpc().addAccount(newPrivateKey);
```

Comment:
This is related to the above comment.  I need to test this more but I might need to figure out a way to get this to wait until the transaction is approved by the user before calling the addAccount function.